### ISSUE TITLE
line of code stranded after exception explicitly raised in IMAP4.client.__cbContinueAuth

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3751]
|<img alt="whysean's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| whysean reported|
|-|-|
|Trac ID|trac#3751|
|Type|defect|
|Created|2009-04-06 03:14:53Z|
|Branch|https://github.com/twisted/twisted/tree/imap4-spurious-loseconn-3751|

```IMAP4.client.__cbContinueAuth``` function has two lines of code that appear to be out of order. These are lines 2455 and 2456 in the file imap4.py 

```python
    def __cbContinueAuth(self, rest, scheme, secret):
        try:
            chal = base64.decodestring(rest + '\n')
        except binascii.Error:
            self.sendLine('*')
--->        raise IllegalServerResponse(rest)
--->        self.transport.loseConnection()
        else:
            auth = self.authenticators[scheme]
            chal = auth.challengeResponse(secret, chal)
            self.sendLine(base64.encodestring(chal).strip())
```

Attachments:

* [mail-auth-3751.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/e93/e9301444226221ab499d9a832cb6c5550c6b7a81/cc7e79aa421a050bfbdae52a9983053673b66cd1.patch) (684 bytes) - added by thijs on 2011-05-27 11:23:38Z - 
* [3751-cbContinueAuth-20120727.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/e93/e9301444226221ab499d9a832cb6c5550c6b7a81/58407e73de82b3968d1f3bdff641c5bc76524d22.patch) (2024 bytes) - added by argonemyth on 2012-07-27 17:22:58Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3751 3751
type__defect defect
reporter__whysean whysean
priority__normal normal
milestone__ 
branch__branches_imap4_spurious_loseconn_3751 branches/imap4-spurious-loseconn-3751
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__ 
time__1238987693000000 1238987693000000
changetime__1348863787000000 1348863787000000
version__None None
owner__exarkun exarkun
cc__thijs cc__argonemyth
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_1"></a>|@thijstriemstra commented|
|-|-|

add markup to ticket description

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_4"></a>|@thijstriemstra commented|
|-|-|

Attached patch switches the lines. I'm not sure how to test this (if it isn't already, tests all still pass) so putting the patch up for review, hopefully it's as simple as applying the patch.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve set owner to @thijstriemstra|
|-|-|

It definitely needs a test.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @argonemyth|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="argonemyth's avatar" src="https://avatars.githubusercontent.com/u/965739?s=50" width="50" height="50">](https://github.com/argonemyth)<a name="note_8"></a>|@argonemyth removed owner|
|-|-|

I noticed that the line **self.transport.loseConnection()** in __cbContinueAuth() is not needed at all. **IMAP4Client.dispatchCommand()** should catch the exception & call **self.transport.loseConnection()**.

I added a test in **HandCraftedTestCase** - not sure if it's the right place to put it. I am not good at naming stuff. Please let me know if you come up with a better function name for test_IMAP4ClientContinueAuthCallbackWithBinasciiError.

Please review, thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [35816]) Branching to 'imap4-spurious-loseconn-3751'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [35817]) Apply 3751-cbContinueAuth-20120727.patch 

refs [#3751](https://github.com/twisted/twisted/issues/3751)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set owner to @exarkun|
|-|-|

Thanks.  Sorry about the delay in the review.  Some comments:

  1. The test case name is a bit overly precise while also including unnecessary words.
    1. All of the tests on this class are for the `IMAP4Client`, so the test method name isn't telling us much by including that.
    1. `ContinueAuthCallback` is reflecting a private implementation detail that we don't really need to care about.  There is public, externally visible behavior this change affects.  The exact shape of the implementation doesn't really matter.
  1. The line that delivers the server greeting text to the protocol is too long - lines need to be 80 columns or narrower.
  1. `_ebCatchLogerr` doesn't prove the `Deferred` returned from `authenticate` fails, it just silences it.  It's pretty important for the `Deferred` to errback in this case, or client applications will never get notified that the connection was closed and no further communication is happening with the server.  `TestCase.assertFailure` is a great helper for testing things like this.
  1. I think the (existing, unchanged) behavior is a little bit amiss.  I filed [#6021](https://github.com/twisted/twisted/issues/6021) for improving it in the future.

Thanks!  I'll make some minor modifications, verify that it works on all platforms, and then apply the change to trunk.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

(In [35818]) address review feedback, add news fragment

refs [#3751](https://github.com/twisted/twisted/issues/3751)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

(In [35819]) Use the proper name for ConnectionDone

refs [#3751](https://github.com/twisted/twisted/issues/3751)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set status to `closed`|
|-|-|

(In [35839]) Merge imap4-spurious-loseconn-3751

Author: argonemyth
Reviewer: exarkun
Fixes: [#3751](https://github.com/twisted/twisted/issues/3751)

Remove an extranous `loseConnection` call from `IMAP4Client` and improve test
coverage of the related code paths.
